### PR TITLE
Add repo test and aggregations

### PR DESCRIPTION
**Pull request by ajax-marchenko-s**
- add mongock
- add aggregation for artwork and auction
- add test for repository

### CODE REVIEW

File: src/test/kotlin/ua/marchenko/artauction/artwork/repository/ArtworkRepositoryTest.kt [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    @Test
+    fun `should save artwork`() {
+        // GIVEN
+        val artwork = MongoArtwork.random(id = null)
+
+        // WHEN
+        val savedArtwork = artworkRepository.save(artwork)
+
+        // THEN
+        assertEquals(artwork.title, savedArtwork.title)
```
Comment (ajax-torbin-v) [inline src/test/kotlin/ua/marchenko/artauction/artwork/repository/ArtworkRepositoryTest.kt:34]: I think it's better here to fully compare objects

File: src/test/kotlin/ua/marchenko/artauction/artwork/repository/ArtworkRepositoryTest.kt [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        assertEquals(savedArtwork.id, result?.id)
+    }
+
+    @Test
+    fun `should return null when there is no artwork with this id`() {
+        // WHEN
+        val result = artworkRepository.findById(getRandomObjectId().toString())
+
+        // THEN
+        assertNull(result, "Found artwork must be null")
```
Comment (ajax-torbin-v) [inline src/test/kotlin/ua/marchenko/artauction/artwork/repository/ArtworkRepositoryTest.kt:55]: // GIVEN 
?

File: src/test/kotlin/ua/marchenko/artauction/artwork/repository/ArtworkRepositoryTest.kt [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        val result = artworkRepository.findById(savedArtwork.id.toString())
+
+        // THEN
+        assertEquals(savedArtwork.id, result?.id)
+    }
+
+    @Test
+    fun `should return null when there is no artwork with this id`() {
+        // WHEN
+        val result = artworkRepository.findById(getRandomObjectId().toString())
```
Comment (ajax-torbin-v) [inline src/test/kotlin/ua/marchenko/artauction/artwork/repository/ArtworkRepositoryTest.kt:52]: ObjectId() will create random object id

File: src/test/kotlin/ua/marchenko/artauction/artwork/repository/ArtworkRepositoryTest.kt [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+        // WHEN
+        val result = artworkRepository.findAll()
+
+        // THEN
+        assertTrue(result.size >= artworks.size, "Size of list must be at least ${result.size}")
+        artworks.forEach { artwork ->
+            assertTrue(result.any { it.title == artwork.title }, "Artwork with title ${artwork.title} must be found")
+        }
+    }
```
Comment (ajax-torbin-v) [inline src/test/kotlin/ua/marchenko/artauction/artwork/repository/ArtworkRepositoryTest.kt:120]: When this test can possibly fail?
Seems like you testing save here

File: src/test/kotlin/ua/marchenko/artauction/artwork/service/ArtworkServiceTest.kt [hunk 5]
```diff
@@ -91,12 +92,51 @@ class ArtworkServiceTest {
...
+
+        // THEN
+        assertEquals(artwork, result)
+    }
+
+    @Test
+    fun `should throw ArtworkNotFoundException when there is no full artwork with this id`() {
+        //GIVEN
+        val id = getRandomObjectId().toHexString()
+        every { mockArtworkRepository.findFullById(id) } returns null
```
Comment (ajax-torbin-v) [inline src/test/kotlin/ua/marchenko/artauction/artwork/service/ArtworkServiceTest.kt:114]:
```suggestion
        every { mockArtworkRepository.findFullById(ObjectId()) } returns null
```

File: src/main/kotlin/ua/marchenko/artauction/artwork/controller/ArtworkController.kt [hunk 6]
```diff
@@ -22,10 +23,18 @@ class ArtworkController(private val artworkService: ArtworkService) {
...
+    @ResponseStatus(HttpStatus.OK)
+    fun getFullArtworkById(@PathVariable id: String) = artworkService.getFullById(id).toArtworkFullResponse()
+
     @GetMapping
     @ResponseStatus(HttpStatus.OK)
     fun getAllArtworks() = artworkService.getAll().map { it.toArtworkResponse() }
 
+    @GetMapping("/full")
+    @ResponseStatus(HttpStatus.OK)
+    fun getAllFullArtworks() = artworkService.getFullAll().map { it.toArtworkFullResponse() }
```
[the remark this answers is not in the document]
Comment (ajax-marchenko-s) [reply src/main/kotlin/ua/marchenko/artauction/artwork/controller/ArtworkController.kt]: add limit and page

File: src/main/kotlin/ua/marchenko/artauction/artwork/repository/impl/MongoArtworkRepository.kt [hunk 7]
```diff
@@ -4,24 +4,59 @@ import org.springframework.data.mongodb.core.MongoTemplate
...
+
+    override fun findAll(): List<MongoArtwork> = mongoTemplate.findAll(MongoArtwork::class.java)
+
+    override fun findFullAll(): List<ArtworkFull> {
+        val aggregation = Aggregation.newAggregation(
+            *aggregateFullArtist().toTypedArray(),
+        )
+        val results = mongoTemplate.aggregate(aggregation, MongoArtwork.COLLECTION, ArtworkFull::class.java)
+        return results.mappedResults.toList()
+    }
```
[the remark this answers is not in the document]
Comment (ajax-marchenko-s) [reply src/main/kotlin/ua/marchenko/artauction/artwork/repository/impl/MongoArtworkRepository.kt]: added skip and limit

File: src/test/kotlin/ua/marchenko/artauction/artwork/repository/ArtworkRepositoryTest.kt (hunk 2 above)
Comment (ajax-marchenko-s) [reply src/test/kotlin/ua/marchenko/artauction/artwork/repository/ArtworkRepositoryTest.kt]: nothing

File: src/test/kotlin/ua/marchenko/artauction/artwork/repository/ArtworkRepositoryTest.kt (hunk 4 above)
Comment (ajax-marchenko-s) [reply src/test/kotlin/ua/marchenko/artauction/artwork/repository/ArtworkRepositoryTest.kt:120]: if saved artwork would not be found in returned list

File: src/test/kotlin/ua/marchenko/artauction/artwork/service/ArtworkServiceTest.kt (hunk 5 above)
Comment (ajax-marchenko-s) [reply src/test/kotlin/ua/marchenko/artauction/artwork/service/ArtworkServiceTest.kt]: changed to any()

File: src/main/kotlin/ua/marchenko/artauction/artwork/model/MongoArtwork.kt [hunk 8]
```diff
@@ -1,15 +1,13 @@
...
 import org.springframework.data.mongodb.core.mapping.Document
 import org.springframework.data.mongodb.core.mapping.MongoId
 import ua.marchenko.artauction.artwork.enums.ArtworkStatus
 import ua.marchenko.artauction.artwork.enums.ArtworkStyle
-import ua.marchenko.artauction.user.model.User
 
-@Document(collection = "artwork")
-data class Artwork(
+@Document(collection = MongoArtwork.COLLECTION)
+data class MongoArtwork(
```
Comment (ajax-samondros-b) [inline src/main/kotlin/ua/marchenko/artauction/artwork/model/MongoArtwork.kt:10]: add type alias always by default

File: src/main/kotlin/ua/marchenko/artauction/auction/controller/dto/AuctionResponse.kt [hunk 9]
```diff
@@ -1,14 +1,18 @@
...
-    val bid: Double,
-    val buyer: UserResponse?,
+    val artworkId: String,
+    val startBid: BigDecimal,
+    val buyers: List<BidResponse>,
     val startedAt: LocalDateTime,
     val finishedAt: LocalDateTime,
-)
+) {
+    data class BidResponse(
```
Comment (ajax-samondros-b) [inline src/main/kotlin/ua/marchenko/artauction/auction/controller/dto/AuctionResponse.kt:14]: More meaningful name

File: src/main/kotlin/ua/marchenko/artauction/auction/mapper/AuctionMapperExtension.kt [hunk 10]
```diff
@@ -1,22 +1,45 @@
...
+    artworkId?.toHexString() ?: "unknown",
+    startBid ?: BigDecimal(0.0),
+    buyers?.map { it.toBidResponse() }.orEmpty(),
     startedAt ?: LocalDateTime.MIN,
     finishedAt ?: LocalDateTime.MIN,
 )
 
-fun CreateAuctionRequest.toAuction(artwork: Artwork, buyer: User?) =
-    Auction(null, artwork, bid, buyer, startedAt, finishedAt)
+fun CreateAuctionRequest.toAuction() =
```
Comment (ajax-samondros-b) [inline src/main/kotlin/ua/marchenko/artauction/auction/mapper/AuctionMapperExtension.kt:25]: toDomain(), toMongo(), toResponse(), etc...

File: src/main/kotlin/ua/marchenko/artauction/common/mongock/migration/ArtAuctionMigrationChangeUnit.kt [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+import ua.marchenko.artauction.user.model.MongoUser
+
+@ChangeUnit(id = "ArtAuctionMigrationChangeUnit", order = "001", author = "svitlana marchenko", systemVersion = "1")
+class ArtAuctionMigrationChangeUnit {
+
+    @Execution
+    fun createIndex(mongoTemplate: MongoTemplate) {
+        val indexOps: IndexOperations = mongoTemplate.indexOps(MongoUser.COLLECTION)
+        indexOps.ensureIndex(
+            Index().on(MongoUser::email.name, Sort.Direction.ASC).unique()
```
Comment (ajax-samondros-b) [inline src/main/kotlin/ua/marchenko/artauction/common/mongock/migration/ArtAuctionMigrationChangeUnit.kt:20]: hashed index is more suitable here

File: src/main/kotlin/ua/marchenko/artauction/common/mongock/migration/ArtAuctionMigrationChangeUnit.kt [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+import io.mongock.api.annotations.RollbackExecution
+import org.slf4j.LoggerFactory
+import org.springframework.data.domain.Sort
+import org.springframework.data.mongodb.core.MongoTemplate
+import org.springframework.data.mongodb.core.index.Index
+import org.springframework.data.mongodb.core.index.IndexOperations
+import ua.marchenko.artauction.user.model.MongoUser
+
+@ChangeUnit(id = "ArtAuctionMigrationChangeUnit", order = "001", author = "svitlana marchenko", systemVersion = "1")
+class ArtAuctionMigrationChangeUnit {
```
Comment (ajax-samondros-b) [inline src/main/kotlin/ua/marchenko/artauction/common/mongock/migration/ArtAuctionMigrationChangeUnit.kt:14]: Why migration is called ArtAuctionMigrationChangeUnit but you create index for user inside. Mirror migration action to its name

**Review by ajax-samondros-b — APPROVED** [review]

**Pull request merged**